### PR TITLE
[fix]: add Arm 4bit fused moe support

### PR DESCRIPTION
**Pull request by nikhil-arm**
## Purpose
Add support for fused 4bit moe for Arm platform
## Test Result
Tested with Deepseek v2 lite model. The generation is correct with upto 3x speedups.
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 1]
```diff
@@ -1639,3 +1644,290 @@ def apply(
...
+                         scales_2d: torch.Tensor,
+                         bias_1d: Optional[torch.Tensor], in_features: int,
+                         out_features: int) -> torch.Tensor:
+            # int4 values are stored as int8 in [-8,7].
+            # Shift to unsigned nibble and pack pairs along input-dim.
+            tmp = int4_as_int8_2d.add(8)  # [out, in]
+            uint8_nibbles = ((tmp[:, 1::2] << 4) | tmp[:, ::2]).to(
+                torch.uint8)  # [out, in//2]
+
+            scales = scales_2d.to(torch.float32)
```
Comment (nikhil-arm) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: fix scale & bias dtype hardcoding

File: csrc/moe/dynamic_4bit_int_moe_cpu.cpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
+#include <ATen/ATen.h>
+#include <ATen/Parallel.h>
+#include <ATen/ops/_dyn_quant_matmul_4bit.h>
+#include <torch/all.h>
+#include <ATen/cpu/vec/vec.h>
```
Comment (nikhil-arm) [inline csrc/moe/dynamic_4bit_int_moe_cpu.cpp]: remove unused headers

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 3]
```diff
@@ -1639,3 +1644,290 @@ def apply(
...
+
+        def _act_kind(s: str) -> int:
+            # 0 = SwiGLU_Gu (SiLU(g)*u), 1 = SwiGLU_Ug (SiLU(u)*g), 2 = SiLU
+            if s == "swiglu":
+                return 0
+            if s == "swigluoai":
+                return 1
+            if s == "silu":
+                return 2
+            raise ValueError(f"Unknown activation '{s}'")
```
[the remark this answers is not in the document]
Comment (nikhil-arm) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py:1970]: vLLM treats silu as siluandmul for moe. swiglu_gu is correct implementation for silu or swiglu

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 4]
```diff
@@ -1639,3 +1644,290 @@ def apply(
...
+        router_logits: torch.Tensor,
+        top_k: int,
+        renormalize: bool,
+        use_grouped_topk: bool = False,
+        topk_group: Optional[int] = None,
+        num_expert_group: Optional[int] = None,
+        global_num_experts: int = -1,
+        expert_map: Optional[torch.Tensor] = None,
+        custom_routing_function: Optional[Callable] = None,
+        scoring_func: str = "softmax",
```
Comment (aditew01) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py:1945]: Note: add `routed_scaling_factor` param to be in sync with latest changes: 
ref: https://github.com/vllm-project/vllm/blob/4f6593b058dc7ba66d887442ba5763c6c1b3886e/vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py#L353

Comment (nikhil-arm) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py:1945]: done

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 1 above)
Comment (nikhil-arm) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: done

**Comment by nikhil-arm** [comment]
@mgoin can we please review and merge this

File: csrc/moe/dynamic_4bit_int_moe_cpu.cpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+          mm(x_e, w13_e, g_eff_13, /*in_features=*/H, /*out_features=*/I2);
+
+      auto g_part = y13.narrow(/*dim=*/1, /*start=*/0, /*length=*/I);
+      auto u_part = y13.narrow(/*dim=*/1, /*start=*/I, /*length=*/I);
+
+      // We Intentionally Club Gate and Up projection for  4 bit MOE
+      // In this case SiLU becomes SwigLU_GU
+      // GptOSS uses Reversed Swiglu even if we concatanate Gate and Up
+      torch::Tensor act;
+      if (activation_kind == ActivationKind::SwiGLU_Ug) {  // SwiGLU_UG
```
Comment (nikhil-arm) [inline csrc/moe/dynamic_4bit_int_moe_cpu.cpp]: use openai specific limits and alpha with sigmoid

File: vllm/model_executor/layers/fused_moe/cpu_fused_moe.py [hunk 6]
```diff
@@ -98,13 +98,19 @@ def select_experts(
                             e_score_correction_bias=e_score_correction_bias)
     elif custom_routing_function is None:
         assert scoring_func == "softmax"
-        topk_weights = torch.nn.functional.softmax(router_logits,
```
Comment (xiangze-arm) [inline vllm/model_executor/layers/fused_moe/cpu_fused_moe.py:101]: This change will break Qwen3 MoE models, which use softmax then topK in router.

File: vllm/model_executor/layers/fused_moe/cpu_fused_moe.py [hunk 7]
```diff
@@ -98,13 +98,19 @@ def select_experts(
...
+                                               k=top_k,
+                                               dim=-1,
+                                               sorted=False)
         if renormalize:
-            topk_weights /= topk_weights.sum(dim=-1, keepdim=True)
-        return topk_weights, topk_ids.to(torch.int32)
+            sel = topk_logit_vals - topk_logit_vals.max(dim=-1,
+                                                        keepdim=True).values
+            topk_vals = sel.exp()
+            topk_vals = topk_vals / topk_vals.sum(dim=-1, keepdim=True)
```
Comment (xiangze-arm) [inline vllm/model_executor/layers/fused_moe/cpu_fused_moe.py]: I think these 4 lines are equivalent to a Softmax?

Comment (nikhil-arm) [reply vllm/model_executor/layers/fused_moe/cpu_fused_moe.py]: for primitive ops, pytorch uses vectorizer classes internally. These vectoriser classes are highly optimised. 
My idea was to use combination of ops instead of .softmax() so that vLLM can alwasy benefit from vectorizer improvements.

I will profile .softmax() vs combo vectorizer op code and update it

Thanks for pointing this out

Comment (nikhil-arm) [reply vllm/model_executor/layers/fused_moe/cpu_fused_moe.py]: torch.softmax() is faster

Comment (nikhil-arm) [reply vllm/model_executor/layers/fused_moe/cpu_fused_moe.py]: done

**Review by mgoin — APPROVED** [review]
LGTM, nice work!

**Comment by mgoin** [comment]
Looks like it failed to compile https://buildkite.com/vllm/ci/builds/31521/steps/canvas?jid=019962aa-5f25-4a05-b3bd-cd4cd2973814#019962aa-5f25-4a05-b3bd-cd4cd2973814/132-5931

**Comment by nikhil-arm** [comment]
> Looks like it failed to compile https://buildkite.com/vllm/ci/builds/31521/steps/canvas?jid=019962aa-5f25-4a05-b3bd-cd4cd2973814#019962aa-5f25-4a05-b3bd-cd4cd2973814/132-5931

Thanks, I have fixed it and pushed the changes
